src/discordcr/cache.cr

```diff
@@ -1,216 +1,241 @@
+      cache(object.id, object)
     end
 
-    # Resolves the ID of a DM channel with a particular user by the recipient's
-    # *recipient_id*. If there is no such channel cached, one will be created.
-    def resolve_dm_channel(recipient_id : UInt64) : UInt64
-      @dm_channels.fetch(recipient_id) do
-        channel = @client.create_dm(recipient_id)
-        cache(Channel.new(channel))
-        channel.id
-      end
-    end
+    # Removes a member from this cache from its `key`.
+    abstract def remove(key : K)
 
-    # Resolves the current user's profile. Requires no parameters since the
-    # endpoint has none either. If there is a gateway connection this should
-    # always be cached.
-    def resolve_current_user : User
-      @current_user ||= @client.get_current_user
+    # Removes a member, expecting that `object` responds to `#id`.
+    def remove(object : V)
+      remove(object.id)
     end
 
-    # Deletes a user from the cache given its *ID*.
-    def delete_user(id : UInt64)
-      @users.delete(id)
-    end
+    # Resolves the cache memeber from `key`.
+    abstract def resolve?(key : K)
 
-    # Deletes a channel from the cache given its *ID*.
-    def delete_channel(id : UInt64)
-      @channels.delete(id)
-    end
+    # Resolves the cache member from `key`.
+    # Raises an exception if the member isn't found instead of making a request.
+    def resolve(key : K)
+      member = resolve?(key)
+      raise MissingMember.new("Cache member not present with key: #{key}") unless member
 
-    # Deletes a guild from the cache given its *ID*.
-    def delete_guild(id : UInt64)
-      @guilds.delete(id)
+      member
     end
 
-    # Deletes a member from the cache given its *user_id* and the *guild_id* it
-    # is on.
-    def delete_member(guild_id : UInt64, user_id : UInt64)
-      @members[guild_id]?.try &.delete(user_id)
-    end
+    # Registers handlers on this cache's `Client` to manage its stored data
+    abstract def init_handlers
+  end
 
-    # Deletes a role from the cache given its *ID*.
-    def delete_role(id : UInt64)
-      @roles.delete(id)
-    end
+  # Container for multiple `TypeCache` objects that maps Discord types to a `TypeCache`
+  class CacheSet
+    # The `Client` to pass to the caches to register their handlers on after initialization
+    getter client : Client
+
+    def initialize(@client)
+    end
+
+    # The different possible cache types
+    TYPES = {
+      guild:        {key: UInt64, type: Guild},
+      guild_member: {key: {UInt64, UInt64}, type: GuildMember},
+      channel:      {key: UInt64, type: Channel},
+      role:         {key: {UInt64, UInt64}, type: Role},
+      users:        {key: UInt64, type: User},
+    }
+
+    {% begin %}
+      {% for key, properties in TYPES %}
+        # Cache for `{{properties[:type]}}` objects
+        property {{key}} : TypeCache({{properties[:key]}}, {{properties[:type]}})?
+
+        def [](type : {{properties[:type]}}.class)
```

Comment:
This method is ambiguous if there are multiple cache types with the same `type` property (e.g. guild-role ID mapping and guild-channel ID mapping)

Reply:
This is intentional ; I wasn't planning on "allowing" multiple stores for a single type. This method is really here to simplify the `cached_route` macro in `CachedClient`, where I can pass the route's argument type instead of transforming the AST to call the `property {{key}}`.

Anyways; are multiple stores of a given type really desired? I can't see why anyone would store things, for example, both in-memory *and* in a database.